niworkflows/interfaces/utils.py

```diff
@@ -77,3 +112,95 @@ def _run_interface(self, runtime):
+    """
+    Generates a reference grid for resampling one image keeping original resolution,
+    but moving data to a different space (e.g. MNI)
+    """
+
+    input_spec = GenerateSamplingReferenceInputSpec
+    output_spec = GenerateSamplingReferenceOutputSpec
+
+    def _run_interface(self, runtime):
+        self._results['out_file'] = _gen_reference(
+            self.inputs.fixed_image,
+            self.inputs.moving_image,
+            force_xform_code=self.inputs.xform_code,
+            message='%s of niworkflows v%s' % (self.__class__.__name__, __version__))
+        return runtime
+
+
+def _copyxform(ref_image, out_image, message=None):
+    # Read in reference and output
+    resampled = nb.load(out_image)
+    orig = nb.load(ref_image)
+
+    if not np.allclose(orig.affine, resampled.affine):
+        LOG.warning('Affines of input and reference images '
+                    'do not match, CopyXForm will probably '
+                    'make the input image useless.')
+
+    # Copy xform infos
+    qform, qform_code = orig.header.get_qform(coded=True)
+    sform, sform_code = orig.header.get_sform(coded=True)
+    header = resampled.header.copy()
+    header.set_qform(qform, int(qform_code))
+    header.set_sform(sform, int(sform_code))
+    header['descrip'] = 'header fixed (%s)' % (message or 'unknown')
+
+    newimg = resampled.__class__(resampled.get_data(), orig.affine, header)
+    newimg.to_filename(out_image)
+
+
+def _gen_reference(fixed_image, moving_image, out_file=None, message=None,
+                   force_xform_code=None):
+    """
+    Generates a sampling reference, and makes sure xform matrices/codes are
+    correct
+    """
+
+    if out_file is None:
+        out_file = fname_presuffix(fixed_image,
+                                   suffix='_reference',
+                                   newpath=os.getcwd())
+
+    new_zooms = nli.load_img(moving_image).header.get_zooms()[:3]
+    # Avoid small differences in reported resolution to cause changes to
+    # FOV. See https://github.com/poldracklab/fmriprep/issues/512
+    new_zooms_round = np.round(new_zooms, 3)
+
+    resampled = nli.resample_img(fixed_image,
+                                 target_affine=np.diag(new_zooms_round),
+                                 interpolation='nearest')
+
+    xform = resampled.affine  # nibabel will pick the best affine
+    _, qform_code = resampled.header.get_qform(coded=True)
+    _, sform_code = resampled.header.get_sform(coded=True)
+
+    xform_code = sform_code if sform_code > 0 else qform_code
+    if xform_code == 1:
+        xform_code = 2
+
+    if force_xform_code is not None:
+        xform_code = force_xform_code
+
+    # Keep 0, 2, 3, 4 unchanged
+    resampled.header.set_qform(xform, xform_code)
+    resampled.header.set_sform(xform, xform_code)
+    resampled.header['descrip'] = 'reference image generated by %s.' % (
+        message or '(unknown software)')
```

Comment:
So all of this... it's just updating the `qform_code` to match the `sform_code`, and setting `descrip` field. Because `resampled` is being generated by nilearn/nibabel, `qform` already `== sform` (modulo shearing):

https://github.com/nipy/nibabel/blob/8c1d0bc4bd9cd757f9b9658547f0e2af55e8ca6c/nibabel/nifti1.py#L1789-L1795

It's not wrong, just I'm not sure what we're buying here.

Reply:
I see. You are right, this is not buying much. I'd prefer to keep it though, to have your code handy when we need it, and still the interface now can manually set a different code, which will be useful when we actually start using it.